peps/pep-0724.rst

```diff
@@ -0,0 +1,305 @@
+         if is_cardinal_direction(direction):
+             # Literal["E"] The type cannot be "N", "S" or "W" here because of
+             # argument type
+         else:
+             # Literal["NW"]
+
+
+This logic can be summed up in a table:
+
+============ ======================= ===================
+\            Non strict TypeGuard    Strict TypeGuard
+============ ======================= ===================
+Applies when R not consistent with P R consistent with P
+NP is ..     :math:`R`               :math:`A \land R`
+NN is ..     :math:`A`               :math:`A \land \neg{R}`
+============ ======================= ===================
+
+Intersection of types and type negation are not defined and are left up to the
+type checker to decide on how to implement. Future extensions to the type
+system may define this behavior though.
+
+
+Additional Examples
+===================
+
+Any
+---
+
+``Any`` is consistent [#isconsistent]_ with any other type, which means
+stricter semantics can be applied.
+
+.. code-block:: python
+
+    def is_a(x: Any) -> TypeGuard[A]:  # Stricter TypeGuard possible
+        return isinstance(x, A)
+
+    def test(x: A | B):
+        if is_a(x):
+            # x is of type A here
+        else:
+            # x is of type B here
+
+
+Backwards Compatibility
+=======================
+
+The new ``False`` case for a ``TypeGuard`` breaks backwards compatibility.
+
+.. code-block:: python
+
+    def is_int(val: int | str) -> TypeGuard[int]:
+        return isinstance(val, int)
+
+    def func(val: int | str):
+        if is_int(val):
+            # Type checker assumes int here
+        else:
+            # Type checker assumes str here
+
+A type checker will assume in the ``False`` case that the value is ``str``.
+
+This is a change in behavior from :pep:`647`.
+
+Although this is different behavior, existing code would likely have further
+checks for a ``str`` in the ``False`` case. Since the previous behavior was to
+assume ``str | int``, existing code would have had to further refine the type
+to get a type checker to behave the expected way.
+
+Something like so:
+
+.. code-block:: python
+
+    def is_int(val: int | str) -> TypeGuard[int]:
+        return isinstance(val, int)
+
+    def is_str(val: int | str) -> TypeGuard[int]:
+        return isinstance(val, int)
+
+    def func(val: int | str):
+        if is_int(val):
+            # Code does something with int here
+        else is_str(val):  # This check would have been necessary before
+            # Code does something with str here
+
+
+This means existing code should continue to work with the ``False`` narrowing.
+
+As a proof of this concept, we ran the `experimental Pyright changes`__
+against `mypy primer`__ to see if there were any differences in the output.
+Changing ``TypeGuard`` to be more strict had a small effect on the ``mypy
+primer``. Mostly indicating that some ``# type: ignore`` comments were no longer
+necessary.
```

Comment:
mypy primer is the implicit object of the sentence here, I think this phrasing is nicer
```suggestion
Changing ``TypeGuard`` to be more strict had a small effect,
mostly indicating that some ``# type: ignore`` comments were no longer
necessary.
```